DIPs/DIPNNNN.md

```diff
@@ -0,0 +1,349 @@
+}
+```
+
+It makes naming clashes even less likely and resulting names more obvious
+but suffers from different important issues:
+
+- Can't be used for new symbols only, fundamental library API design decision
+- Does not leverage D module system
+- Can't be use to solve [Problem 2](#problem-2-override-clash)
+
+#### Deprecations
+
+Contrary to adding new symbol, removing or renaming one is much easier task
+because D supports `deprecated` keyword, ensuring that users of the library
+will get early notification about to-be-removed symbol before it is actually
+gone.
+
+Similar warning should also be suitable for adding new symbols in critical
+libraries but language currently lack a way to express such semantics.
+
+### Proposal
+
+Introduce new symbol kind concept into compiler, *future symbol*. In terms of
+DMD compiler internals it can be defined simply as additional boolean field
+of `DSymbol` indicating if symbol is *future* or not.
+
+To make examples more readable, imaginary `@future` attribute will be used to
+mark such symbols. However introducing such attribute (or any other) is NOT
+part of the proposal and serves only illustration purpose.
+
+Symbol marked as *future* has to have special treatment during semantic analysis:
+
+1. Whenever compiler has to do a symbol lookup, if has to check if *future
+   symbol* is present in the result. If present, deprecation message has to be
+   printed and everything else should continue as if it wasn't present.
+
+   ```D
+   import moda; // provides foo
+   import modb; // provides @future foo
+
+   foo(); // Deprecation: upcoming adddition of modb.foo will result in a symbol
+          // clash, please disambugate
+   ```
+
+2. Whenevr compiler hasto check for presence of a symbol, if there is a match
+   which is *future symbol*, deprecation message has to be printed and
+   everything else should continue as it wasn't present.
+
+   ```D
+   class A
+   {
+       void foo() @future;
+   }
+
+   class B
+   {
+       void foo(); // Deprecation: upcoming addition of A.foo will result in a
+                   // symbol clash, please adjust
+   }
+   ```
+
+3. References to *future symbol* within the module it is declared in should not
+   result in any deprecation messages because the module can be adjusted in the
+   same time symbol is actually added.
+
+   ```D
+   void foo (long a);
+   void foo (int a) @future;
+
+   foo(42); // No point in printing deprecation here as it will keep compiling
+            // even when @future is removed
+   ```
+
+4. Any form of static reflection that provides sequence of results, for example
+   `__traits(allMembers)` or `.tupleof`, has to ignore any *future symbol*
+   completely. Such code does not normally rely on presence of any specific
+   symbol and thus addition of new one is unlikely to cause problems. On the
+   other hand, issuing deprecation in such context will result in a warning that
+   cannot be addressed by adjusting the code.
```

Comment:
s/issuing deprecation in such context will result in a warning that cannot be addressed by adjusting the code/deprecation warnings issued from within such contexts cannot be addressed by adjusting client code/